Statically linked ockam command binaries

Ockam Command binaries that we produce in this [workflow](https://github.com/build-trust/ockam/blob/develop/.github/workflows/release-binaries.yml#L184-L264), don't currently work in a scratch docker container or alpine containers since they dynamically link to some C dependencies.

The musl rustc targets statically link by default. Maybe we should default to musl
https://doc.rust-lang.org/reference/linkage.html#static-and-dynamic-c-runtimes

I've had a look into that issue, and it seems that static linking using glibc is not really working out due to:
```
  = note: /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libsysinfo-a604d238df1ecc96.rlib(sysinfo-a604d238df1ecc96.sysinfo.573457b8-cgu.13.rcgu.o): in function `core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut':
          sysinfo.573457b8-cgu.13:(.text._ZN4core3ops8function5impls79_$LT$impl$u20$core..ops..function..FnMut$LT$A$GT$$u20$for$u20$$RF$mut$u20$F$GT$8call_mut17h920ca59999c1b5ddE+0x3b5): warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libnix-4ce9b9f6183ed7d2.rlib(nix-4ce9b9f6183ed7d2.nix.0e8a9144-cgu.11.rcgu.o): in function `nix::unistd::initgroups':
          nix.0e8a9144-cgu.11:(.text._ZN3nix6unistd10initgroups17hdb8602f8dad105b5E+0x5): warning: Using 'initgroups' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libsysinfo-a604d238df1ecc96.rlib(sysinfo-a604d238df1ecc96.sysinfo.573457b8-cgu.13.rcgu.o): in function `core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut':
          sysinfo.573457b8-cgu.13:(.text._ZN4core3ops8function5impls79_$LT$impl$u20$core..ops..function..FnMut$LT$A$GT$$u20$for$u20$$RF$mut$u20$F$GT$8call_mut17h84ec919025a3a2feE+0x12): warning: Using 'getgrgid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libnix-4ce9b9f6183ed7d2.rlib(nix-4ce9b9f6183ed7d2.nix.0e8a9144-cgu.11.rcgu.o): in function `nix::unistd::Group::from_gid':
          nix.0e8a9144-cgu.11:(.text._ZN3nix6unistd5Group8from_gid17h7d3d02c84b4ba227E+0xd2): warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libnix-4ce9b9f6183ed7d2.rlib(nix-4ce9b9f6183ed7d2.nix.0e8a9144-cgu.11.rcgu.o): in function `nix::unistd::Group::from_name':
          nix.0e8a9144-cgu.11:(.text._ZN3nix6unistd5Group9from_name17h6a88fd2747c7dbe1E+0x106): warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libnix-4ce9b9f6183ed7d2.rlib(nix-4ce9b9f6183ed7d2.nix.0e8a9144-cgu.11.rcgu.o): in function `nix::unistd::User::from_name':
          nix.0e8a9144-cgu.11:(.text._ZN3nix6unistd4User9from_name17hdf298b6c5ef9a17aE+0x116): warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libnix-4ce9b9f6183ed7d2.rlib(nix-4ce9b9f6183ed7d2.nix.0e8a9144-cgu.11.rcgu.o): in function `nix::unistd::User::from_uid':
          nix.0e8a9144-cgu.11:(.text._ZN3nix6unistd4User8from_uid17h1dd72b8c97696cd6E+0xd2): warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-8f1929c73c3f8167.rlib(std-8f1929c73c3f8167.std.4fd68eda-cgu.0.rcgu.o): in function `<std::sys_common::net::LookupHost as core::convert::TryFrom<(&str,u16)>>::try_from':
          /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/net.rs:205: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
          /usr/bin/ld: /home/hargut/Sources/ockam/target/x86_64-unknown-linux-gnu/release/deps/libockam_command-27ed9d2d09423167.rlib(ockam_command-27ed9d2d09423167.ockam_command.c16c2a85-cgu.13.rcgu.o): undefined reference to symbol '__tls_get_addr@@GLIBC_2.3'
          /usr/bin/ld: /usr/lib64/ld-linux-x86-64.so.2: error adding symbols: DSO missing from command line
          collect2: error: ld returned 1 exit status

```
Which is not easy to resolve as glibc FAQ documents that the option available (--enable-static-nss) would require rebuilding glibc and is not recommended:
https://sourceware.org/glibc/wiki/FAQ#Even_statically_linked_programs_need_some_shared_libraries_which_is_not_acceptable_for_me.__What_can_I_do.3F

Using musl as default library for rust has some voices that report degraded performance compared to glibc.
https://andygrove.io/2020/05/why-musl-extremely-slow/
https://github.com/rust-lang/rust/issues/70108
https://github.com/BurntSushi/ripgrep/issues/1268


Compiling using musl works as expected using:
```
export RUSTFLAGS='-C target-feature=+crt-static'
export PKG_CONFIG_SYSROOT_DIR='/'
cargo build --release --target=x86_64-unknown-linux-musl --release --verbose
```
According to ldd the result has a few libraries linked dynamically including glibc!?
```
[hargut@thinkpad ockam]$ ldd target/x86_64-unknown-linux-musl/release/ockam
        linux-vdso.so.1 (0x00007ffe3c9b3000)
        libssl.so.1.1 => /lib64/libssl.so.1.1 (0x00007f26b3563000)
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007f26b3200000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f26b2e00000)
        libz.so.1 => /lib64/libz.so.1 (0x00007f26b46c8000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f26b46fd000)
[hargut@thinkpad ockam]$
```

--- a/.github/workflows/release-binaries.yml
+++ b/.github/workflows/release-binaries.yml
@@ -196,17 +196,17 @@ jobs:
         - build: linux
           os: ubuntu-18.04
           rust: stable
-          target: aarch64-unknown-linux-gnu
+          target: aarch64-unknown-linux-musl
           use-cross-build: true
         - build: linux_armv7
           os: ubuntu-18.04
           rust: stable
-          target: armv7-unknown-linux-gnueabihf
+          target: armv7-unknown-linux-musleabihf
           use-cross-build: true
         - build: linux_86
           os: ubuntu-18.04
           rust: stable
-          target: x86_64-unknown-linux-gnu
+          target: x86_64-unknown-linux-musl
           use-cross-build: false
         - build: macos
           os: macos-latest
@@ -246,7 +246,7 @@ jobs:
           cargo install --version 0.1.16 cross
         else
           sudo apt-get update
-          sudo apt-get install -y --no-install-recommends xz-utils liblz4-tool
+          sudo apt-get install -y --no-install-recommends xz-utils liblz4-tool musl-tools
         fi
 
     ## TODO Cache for faster build for different targets
@@ -256,6 +256,10 @@ jobs:
       run: |
         set -x
         use_cross_build="${{ matrix.use-cross-build }}"
+
+        if [[ "${{ matrix.target }}" =~ .+-musl(.+)?  ]]; then
+          export RUSTFLAGS='-C target-feature=+crt-static'
+        fi
 
         if [[ $use_cross_build == "true" ]]; then
           cross build --bin ockam --target ${{ matrix.target }} --verbose --release

--- a/Cross.Dockerfile.aarch64
+++ b/Cross.Dockerfile.aarch64
@@ -1,0 +1,5 @@
+FROM ghcr.io/cross-rs/aarch64-unknown-linux-musl:0.2.4
+
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && \
+    apt-get install --assume-yes libdbus-1-dev

--- a/Cross.Dockerfile.armv7
+++ b/Cross.Dockerfile.armv7
@@ -1,0 +1,5 @@
+FROM ghcr.io/cross-rs/armv7-unknown-linux-musleabihf:0.2.4
+
+ARG DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && \
+    apt-get install --assume-yes libdbus-1-dev

--- a/Cross.toml
+++ b/Cross.toml
@@ -1,0 +1,5 @@
+[target.aarch64-unknown-linux-musl]
+dockerfile = "./Cross.Dockerfile.aarch64"
+
+[target.armv7-unknown-linux-musleabihf]
+dockerfile = "./Cross.Dockerfile.armv7"

--- a/deny.toml
+++ b/deny.toml
@@ -16,6 +16,23 @@ allow = [
     "ISC",
     "BlueOak-1.0.0"
 ]
+exceptions = [
+    { allow = ["LicenseRef-Ring-ISC-OpenSSL"], name = "ring", version = "*" },
+]
+
+[[licenses.clarify]]
+name = "ring"
+expression = "LicenseRef-Ring-ISC-OpenSSL"
+license-files = [
+    { path = "LICENSE", hash = 0xbd0eed23 },
+]
+
+[[licenses.clarify]]
+name = "webpki"
+expression = "ISC"
+license-files = [
+    { path = "LICENSE", hash = 0x001c7e6c },
+]
 
 [advisories]
 unmaintained = "deny"

--- a/implementations/rust/ockam/ockam_command/Cargo.toml
+++ b/implementations/rust/ockam/ockam_command/Cargo.toml
@@ -62,7 +62,7 @@ minicbor = { version = "0.18.0", features = ["derive", "alloc", "half"] }
 nix = "0.24"
 open = "2"
 rand = "0.8"
-reqwest = { version = "0.11", default-features = false, features = ["json", "default-tls"] }
+reqwest = { version = "0.11", default-features = false, features = ["json", "rustls-tls-native-roots"] }
 serde = { version = "1", features = ["derive"] }
 serde_json = "1"
 slug = "0.1"

